src/postgres-util/src/lib.rs

```diff
@@ -309,50 +317,88 @@ impl Config {
                 Ok(client)
             }
             SshTunnelConfig::Tunnel { .. } => {
+                let (session, local_port) = self.establish_openssh_connection().await?;
                 let tls = MakeTlsConnect::<TokioTcpStream>::make_tls_connect(&mut tls, &self.host)?;
-                let channel = self.establish_ssh_connection().await?;
+                let tcp_stream = TokioTcpStream::connect(("localhost", local_port)).await?;
                 let (client, connection) = self
                     .postgres_config
                     .replication_mode(ReplicationMode::Logical)
                     .connect_timeout(Duration::from_secs(30))
                     .keepalives_idle(Duration::from_secs(10 * 60))
-                    .connect_raw(channel, tls)
+                    .connect_raw(tcp_stream, tls)
                     .await?;
-                task::spawn(|| "postgres_connect_replication", connection);
+                task::spawn(|| "postgres_connect_replication", async {
+                    _ = connection.await;
+                    _ = session.close().await;
+                });
```

Comment:
can you add a `TODO` to merge this with the above one? I can figure out some spicy way to do it later!

Reply:
You mean the `connect_replication` function with the `connect` function or the `Tunnel` branch with the `Direct` branch?

Reply:
the `connect_replication` and `connect` function!